### ISSUE TITLE
onnx export image size issue

### ISSUE THREAD

**Issue by pranavteli** [ultralytics/ultralytics#10370]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have trained yolov8s segmentation model with imgsz of 1280, when I am exporting the model to onnx, with default imgsz(640), its performing well but when I am exporting with the image size I trained on (1280), its showing poor predictions. Also, segmentation outputsize is 4 times less than imgsz which is resulting in lower resolution masks. Is there any way to tackle this issue. 

Additionally, I want to use segmentation model inference in C# (dot net), I could see only onnxruntime implementation everywhere for segmentation task even official ultralytics mentions converting the model to onnx, Is there any way to directly use pytorch models in C#. I see pytorch implementation i.e. torchsharp but didnt see any example for any computer vision models.

**Comment by glenn-jocher** [comment]
Hello! 🙌

It sounds like you're experiencing a couple of issues related to exporting and using the YOLOv8 segmentation model. Let's address each one:

1. **ONNX Export with Different Image Sizes:** When exporting your model with an image size different from what it was trained on, it's important to ensure the input size during inference matches the export size. Discrepancies in size can lead to poor predictions due to scaling differences. If you've trained your model at a resolution of 1280, try exporting it with the same image size:

   ```bash
   yolo export model=yolov8s-seg.pt format=onnx imgsz=1280,1280
   ```

2. **Segmentation Output Size:** The segmentation masks having a lower resolution is expected behavior due to the nature of downsampling within the network architecture. To get higher resolution masks, you might consider post-processing steps like upsampling the masks to your desired resolution.

3. **Using PyTorch Models in C# (.NET):** As for integrating the segmentation model with C#, while the most straightforward approach is indeed to export the model to ONNX and use it with ONNX Runtime, direct usage of PyTorch models in C# is less common and less documented. TorchSharp is a promising avenue but as you've noticed, examples for computer vision models are scarce. It might require setting up the model architecture manually in C# using TorchSharp and then loading the weights from your trained PyTorch model. This approach, however, can be quite complex and documentation-specific to TorchSharp would be your best reference.

I hope this helps clarify things a bit! If you're tackling the TorchSharp route, it might be worthwhile to reach out in their community forums or documentation for more specific guidance. Good luck with your project! 👍

**Comment by pranavteli** [comment]
> Hello! 🙌
> 
> It sounds like you're experiencing a couple of issues related to exporting and using the YOLOv8 segmentation model. Let's address each one:
> 
> 1. **ONNX Export with Different Image Sizes:** When exporting your model with an image size different from what it was trained on, it's important to ensure the input size during inference matches the export size. Discrepancies in size can lead to poor predictions due to scaling differences. If you've trained your model at a resolution of 1280, try exporting it with the same image size:
>    ```shell
>    yolo export model=yolov8s-seg.pt format=onnx imgsz=1280,1280
>    ```
> 2. **Segmentation Output Size:** The segmentation masks having a lower resolution is expected behavior due to the nature of downsampling within the network architecture. To get higher resolution masks, you might consider post-processing steps like upsampling the masks to your desired resolution.
> 3. **Using PyTorch Models in C# (.NET):** As for integrating the segmentation model with C#, while the most straightforward approach is indeed to export the model to ONNX and use it with ONNX Runtime, direct usage of PyTorch models in C# is less common and less documented. TorchSharp is a promising avenue but as you've noticed, examples for computer vision models are scarce. It might require setting up the model architecture manually in C# using TorchSharp and then loading the weights from your trained PyTorch model. This approach, however, can be quite complex and documentation-specific to TorchSharp would be your best reference.
> 
> I hope this helps clarify things a bit! If you're tackling the TorchSharp route, it might be worthwhile to reach out in their community forums or documentation for more specific guidance. Good luck with your project! 👍

Thank you for taking time to answer this. However, the issue is, I trained model with imgsz=1280, when I am exporting it with default size, its performing well.
`yolo export model=yolov8s-seg.pt format=onnx`
But when I am exporting it to the size I trained on its performing worse.
`yolo export model=yolov8s-seg.pt format=onnx imgsz=1280,1280`

Another interesting fact is- its particularly performing worse for class where object is large. Somehow, its dividing that class objects in multiple instances. This issue is only appearing for large imgsz.

Also, is there any way to modify onnx export architecture itself to add upsampling for segmentation map, so that it has same size as input imgsz, can you please guide me in this case.

**Comment by pranavteli** [comment]
Finally, I could make this work as expected. Sharing code block if anyone come across similar issue.

```
import onnx
from onnx import helper, TensorProto

# Load the original ONNX model
model_path = '/modelPath'
model = onnx.load(model_path)

# Define the upsampling layer
upsample_layer = helper.make_node(
    'Resize',
    inputs=['output1', '', 'scales'],  # Empty string for 'roi'
    outputs=['upsampled_segmaps'],
    mode='linear',  # interpolation
    name='Resize'
)

# scales input for Resize, assuming the desired output size is 4 times the input
scales = helper.make_tensor('scales', TensorProto.FLOAT, [4], [1.0, 1.0, 4.0, 4.0])

# Add nodes and tensors to the graph
nodes = list(model.graph.node) + [upsample_layer]
initializers = list(model.graph.initializer) + [scales]

# Define the outputs, keeping output0 and adding the new upsampled_output1
outputs = [
    model.graph.output[0],  # Keep the original output0
    helper.make_tensor_value_info('upsampled_segmaps', TensorProto.FLOAT, [1, 32, 640, 640]) # Adjust based on your model
]

# Create the modified graph
graph = helper.make_graph(
    nodes=nodes,
    name=model.graph.name,
    inputs=model.graph.input,
    outputs=outputs,
    initializer=initializers,
    value_info=model.graph.value_info
)

# Create the modified model
modified_model = helper.make_model(graph)

modified_model.opset_import[0].version = 17  # Ensure the opset version is correct
modified_model.ir_version = 8

# Save the modified model
onnx.save(modified_model, '/outputPath')
```

**Comment by glenn-jocher** [comment]
Hey, that's fantastic news! 🎉 Thanks for sharing your solution for adding an upsampling layer to the ONNX model. This will definitely be helpful for others facing similar challenges. Your approach to modifying the graph directly is neat and clear. Great job and happy coding! 🚀
